### PR TITLE
Bp 0472 child events

### PR DESCRIPTION
**Pull request by Johannes-Schneider**

### CODE REVIEW

File: src/main/java/de/hpi/bpt/argos/api/entity/EntityEndpointImpl.java [hunk 1]
```diff
@@ -1 +1,349 @@
...
+        JsonObject jsonEventType = getEntityJson(entity, attributes);
+
+        return serializer.toJson(jsonEventType);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public String getChildEntities(Request request, Response response) {
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/api/entity/EntityEndpointImpl.java:91]: MAJOR Line has trailing spaces. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=checkstyle%3ARegexp_Singleline_BP2016)

File: src/main/java/de/hpi/bpt/argos/api/entity/EntityEndpointImpl.java [hunk 2]
```diff
@@ -1 +1,349 @@
...
+     */
+    private JsonArray getEventsJson(List<Event> events) {
+        JsonArray eventsJson = new JsonArray();
+        for (Event event : events) {
+            List<Attribute> attributes = PersistenceAdapterImpl.getInstance().getAttributes(event.getId());
+            JsonArray jsonAttributes = getAttributesJson(attributes);
+
+            JsonObject eventJson = new JsonObject();
+            eventJson.add("Attributes", jsonAttributes);
+
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/api/entity/EntityEndpointImpl.java:256]: BLOCKER NullPointerException might be thrown as 'entityNode' is nullable here [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java [hunk 3]
```diff
@@ -138,7 +138,7 @@ public String executeRequest(Logger logger, Request request, Response response,
 	 */
 	@Override
 	public void logReceivedRequest(Logger logger, Request request) {
-		logger.info(String.format("received request: '%1$s' -> %2$d bytes of content", request.uri(), request.contentLength()));
+		logger.info(String.format("%1$s\t->\t'%2$s' -> %3$d bytes of content", request.ip().toString(), request.uri(), request.contentLength()));
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java]: MAJOR "ip" returns a string, there's no need to call "toString()". [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1858)

File: src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java [hunk 4]
```diff
@@ -153,7 +153,8 @@ public void logSendingResponse(Logger logger, Request request, int responseStatu
 			message = responseMessage;
 		}
 
-		logger.info(String.format("sending response: '%1$s' -> %2$d -> %3$d bytes of response",
+		logger.info(String.format("%1$s\t<-\t'%2$s' -> %3$d -> %4$d bytes of response",
+				request.ip().toString(),
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/util/RestEndpointUtilImpl.java]: MAJOR "ip" returns a string, there's no need to call "toString()". [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1858)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java [hunk 5]
```diff
@@ -161,4 +163,24 @@ public boolean deleteArtifacts(PersistenceArtifact... artifacts) {
 			session.close();
 		}
 	}
+
+	/**
+	 * {@inheritDoc}
+	 */
+	@Override
+	public <ResultType, ResultImplType extends ResultType> List<ResultType> getArtifactsById(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccessImpl.java:171]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccess.java [hunk 6]
```diff
@@ -51,11 +52,28 @@
...
+	/**
+	 * This method returns a list of entities, which will get fetched by their ids.
+	 * @param <ResultType> - the result type
+	 * @param <ResultImplType> - the implementation type of the result type
+	 * @param session - the database session, which must be open
+	 * @param resultTypeClass - the class of the result type
+	 * @param ids - the ids to fetch from the database
+	 * @return - a list of all entities, which were contained in the ids
+	 */
+	<ResultType, ResultImplType extends ResultType> List<ResultType> getArtifactsById(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccess.java:74]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

File: src/main/java/de/hpi/bpt/argos/storage/DatabaseAccess.java [hunk 7]
```diff
@@ -51,11 +52,28 @@
...
 	 * @param getValue - the function to get the results from the query
 	 * @param defaultValue - a fall back default value in case anything went wrong or no entities were found
-	 * @param <R> - the result type
-	 * @param <Q> - the query type
+	 * @param <ResultType> - the result type
+	 * @param <QueryType> - the query type
 	 * @return - an object of the result value type
 	 */
-	<R, Q> R getArtifacts(Session session, Query<Q> query, Transaction transaction, Callable<R> getValue, R defaultValue);
+	<ResultType, QueryType> ResultType getArtifacts(Session session,
```
Comment (hpibptlab) [inline src/main/java/de/hpi/bpt/argos/storage/DatabaseAccess.java:59]: MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)
MINOR Rename this generic name to match the regular expression '^[A-Z][0-9]?$'. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS00119)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 11 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 3 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [PersistenceAdapterImpl.java#L369](https://github.com/bptlab/argos-backend/blob/6cdac2c6f5e2867ba7322ceb63e5688770e35c82/src/main/java/de/hpi/bpt/argos/storage/PersistenceAdapterImpl.java#L369): Define a constant instead of duplicating this literal "eventTypeId" 3 times. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by hpibptlab** [comment]
SonarQube analysis reported 9 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [PersistenceAdapterImpl.java#L369](https://github.com/bptlab/argos-backend/blob/40db5e709cb16a36f35b9b1758858f14e2543a35/src/main/java/de/hpi/bpt/argos/storage/PersistenceAdapterImpl.java#L369): Define a constant instead of duplicating this literal "eventTypeId" 3 times. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Comment by hpibptlab** [comment]
SonarQube analysis reported 9 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [PersistenceAdapterImpl.java#L369](https://github.com/bptlab/argos-backend/blob/c23c4e87cd01960126d0dd0de905d5d73fc2194c/src/main/java/de/hpi/bpt/argos/storage/PersistenceAdapterImpl.java#L369): Define a constant instead of duplicating this literal "eventTypeId" 3 times. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

File: src/test/java/de/hpi/bpt/argos/api/EntityEndpointTest.java [hunk 8]
```diff
@@ -207,14 +208,50 @@ public void testGetEventsOfEntity_SwappedIndices_Success() {
...
+		JsonArray events = jsonParser.parse(request.getResponse()).getAsJsonArray();
+		assertEquals(2, events.size());
+
+		JsonObject event = events.get(0).getAsJsonObject();
+		JsonArray eventAttributes = event.get("Attributes").getAsJsonArray();
+		assertEquals(testEventAttributes.size(), eventAttributes.size());
+
+		event = events.get(1).getAsJsonObject();
+		eventAttributes = event.get("Attributes").getAsJsonArray();
+		assertEquals(newEventAttributes.size(), eventAttributes.size());
```
Comment (frederike-ramin) [inline src/test/java/de/hpi/bpt/argos/api/EntityEndpointTest.java]: How do you make sure the events are different? Those could as well be the attributes of the original testEventType.

**Review by frederike-ramin — CHANGES_REQUESTED** [review]

**Comment by hpibptlab** [comment]
SonarQube analysis reported 9 issues
* ![BLOCKER][BLOCKER] 1 blocker
* ![MAJOR][MAJOR] 1 major
* ![MINOR][MINOR] 7 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![MINOR][MINOR] [PersistenceAdapterImpl.java#L369](https://github.com/bptlab/argos-backend/blob/e8fa5eb8b66d57e5a7c33036ac0299cbb0def886/src/main/java/de/hpi/bpt/argos/storage/PersistenceAdapterImpl.java#L369): Define a constant instead of duplicating this literal "eventTypeId" 3 times. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=squid%3AS1192)

[BLOCKER]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png 'Severity: BLOCKER'
[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'
[MINOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-minor.png 'Severity: MINOR'

**Review by frederike-ramin — APPROVED** [review]

**Pull request merged**